Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h

```diff
@@ -0,0 +1,603 @@
+      { get_local_map().template mark_cell<1>(dh_face, edge_deleted); }
+      if (!get_local_map().is_marked(dh_face, face_deleted))
+      { get_local_map().template mark_cell<2>(dh_face, face_deleted); }
+      Dart_handle dh_adj_face=get_local_map().opposite2(dh_face);
+      if (!get_local_map().is_marked(dh_adj_face, face_deleted))
+      {
+        Dart_handle dh_only_edge=nullptr;
+        if (is_degree_one_face(dh_adj_face, dh_only_edge, edge_deleted))
+        {
+          degree_one_faces.push(dh_only_edge);
+          CGAL_assertion(!get_local_map().is_marked(dh_only_edge, edge_deleted));
+          CGAL_assertion(!get_local_map().is_marked(dh_only_edge, face_deleted));
+        }
+      }
+    }
+    for (auto it=get_local_map().darts().begin(),
+           itend=get_local_map().darts().end(); it!=itend; ++it)
+    {
+      if (it<get_local_map().opposite2(it) && vertex_info(it)>=0 &&
+          !get_local_map().is_marked(it, edge_deleted))
+      { noncon_edges.push_back(it); }
+
+      get_local_map().unmark(it, tested);
+      get_local_map().unmark(it, edge_deleted);
+      get_local_map().unmark(it, face_deleted);
+    }
+
+    get_local_map().free_mark(tested);
+    get_local_map().free_mark(edge_deleted);
+    get_local_map().free_mark(face_deleted);
+  }
+
+  Dart_handle nonhole_dart_of_same_edge(Dart_handle dh)
+  {
+    CGAL_assertion(dh!=nullptr);
+    if (get_local_map().is_marked(dh, m_is_perforated))
+    { dh = get_local_map().opposite2(dh); }
+    CGAL_assertion(!get_local_map().is_marked(dh, m_is_perforated));
+    return dh;
+  }
+
+  Dart_handle nonhole_dart_of_same_edge(Dart_handle dh, bool& flip)
+  {
+    CGAL_assertion(dh!=nullptr);
+    if (get_local_map().is_marked(dh, m_is_perforated))
+    {
+      dh=get_local_map().opposite2(dh);
+      flip=!flip;
+    }
+    CGAL_assertion(!get_local_map().is_marked(dh, m_is_perforated));
+    return dh;
+  }
+
+  void add_to_cycle(Dart_handle dh, Path& cycle, bool flip)
+  {
+    dh=nonhole_dart_of_same_edge(dh, flip);
+    Original_dart_const_handle
+        dh_original=Get_original_dart<Self, Copy>::run(this, dh);
+    if (cycle.can_be_pushed(dh_original, flip))
+    { cycle.push_back(dh_original, flip, false); }
+    else
+    {
+      CGAL_assertion(cycle.can_be_pushed(dh_original, !flip));
+      cycle.push_back(dh_original, !flip, false);
+    }
+  }
+
+  template <class WeightFunctor>
+  bool compute_cycle(Dart_handle root, Path& cycle,
+                     typename WeightFunctor::Weight_t* length,
+                     const typename WeightFunctor::Weight_t* max_length,
+                     const WeightFunctor& wf)
+  {
+    std::vector<typename WeightFunctor::Weight_t> distance_from_root;
+    m_spanning_tree.clear();
+    m_trace_index.clear();
+    initialize_vertex_info();
+    compute_spanning_tree(root, m_spanning_tree, distance_from_root, m_trace_index, wf);
+    compute_noncon_edges(m_spanning_tree, m_noncon_edges);
```

Comment:
I guess `compute_noncon_edges` is not public?

Reply:
yes.